packages/webpack-cli/src/plugins/dotenv-webpack-plugin.js

```diff
@@ -0,0 +1,172 @@
+        vars[key] = value;
+      }
+    });
+
+    return vars;
+  }
+
+  initializeVars() {
+    return this.config.systemvars ? Object.assign({}, process.env) : {};
+  }
+
+  getEnvs(inputFileSystem) {
+    const { path } = this.config;
+
+    const env = dotenv.parse(
+      this.loadFile(path, inputFileSystem),
+      this.getDefaults(inputFileSystem),
+    );
+
+    let blueprint = env;
+
+    return {
+      env,
+      blueprint,
+    };
+  }
+
+  getDefaults(inputFileSystem) {
+    const { path, defaults } = this.config;
+    const defaultFile = defaults === true ? `${path}.defaults` : defaults;
+    return this.loadFile(defaultFile, inputFileSystem);
+  }
+
+  formatData({ variables = {}, target, version }) {
+    const { expand, prefixes } = this.config;
+
+    // add a nomencalture like VITE_VAR_NAME to segregate between the imports
+
+    const preprocessedVariables = Object.keys(variables).reduce((obj, key) => {
+      let value = variables[key];
+      if (expand) {
+        if (value.startsWith("\\$")) {
+          value = value.substring(1);
+        } else if (value.includes("\\$")) {
+          value = value.replace(/\\\$/g, "$");
+        } else {
+          value = interpolate(value, variables);
+        }
+      }
+      obj[key] = JSON.stringify(value);
+      return obj;
+    }, {});
+
+    const formatted = {};
+    prefixes.forEach((prefix) => {
+      Object.entries(preprocessedVariables).forEach(([key, value]) => {
+        formatted[`${prefix}${key}`] = value;
+      });
+    });
+
+    const shouldStubEnv =
+      prefixes.includes("process.env.") &&
+      this.shouldStub({ target, version, prefix: "process.env." });
+    if (shouldStubEnv) {
+      formatted["process.env"] = '"MISSING_ENV_VAR"';
+    }
+
+    return formatted;
+  }
+
+  shouldStub({ target: targetInput, version, prefix }) {
+    if (!version.startsWith("5")) {
+      return false;
+    }
+
+    const targets = Array.isArray(targetInput) ? targetInput : [targetInput];
+
+    return targets.every(
+      (target) =>
+        // If configured prefix is 'process.env'
+        prefix === "process.env." &&
+        // If we're not configured to never stub
+        this.config.ignoreStub !== true &&
+        // And
+        // We are configured to always stub
+        (this.config.ignoreStub === false ||
+          // Or if we should according to the target
+          (!target.includes("node") && !isMainThreadElectron(target))),
+    );
+  }
+
+  /**
+   * Load a file.
+   * @param {String} config.file - The file to load.
+   * @returns {Object}
+   */
+  loadFile(filePath, inputFileSystem) {
+    if (this.cache[filePath]) {
+      return this.cache[filePath];
+    }
+    try {
+      const content = inputFileSystem.readFileSync(filePath, "utf8");
+      this.cache[filePath] = content;
+      return content;
+    } catch (err) {
+      return {};
+    }
+  }
+
+  /**
+   * @param {String} msg - The message.
+   */
+  warn(msg) {
+    console.warn(msg);
```

Comment:
Avoid using `console` in favor `const logger = compilation.getLogger("copy-webpack-plugin");`, let's use webpack logger